### PR TITLE
Constructing nested models for dartsim

### PR DESCRIPTION
**Pull request by brawner**
This PR is ready for review.

Work towards https://github.com/ignitionrobotics/ign-gazebo/issues/479

### CODE REVIEW

File: dartsim/src/SDFFeatures.cc [hunk 1]
```diff
@@ -117,7 +118,6 @@ static void CopyStandardJointAxisProperties(
     const int _index, Properties &_properties,
     const ::sdf::JointAxis *_sdfAxis)
 {
-  _properties.mInitialPositions[_index] = _sdfAxis->InitialPosition();
```
Comment (scpeters) [inline dartsim/src/SDFFeatures.cc:120]: the `//joint/initial_position` element has been deprecated in SDFormat 1.8, but not removed, so I'm inclined to keep this functioning. I would add deprecation warning suppression macros around this call and keep the tests in place. (Those macros don't exist yet in ign-cmake or ign-utils, so I'll work on adding them first

Comment (chapulina) [reply dartsim/src/SDFFeatures.cc:120]: See 35ce9275e3a60077dab35a5a61f8050478abac72 for the deprecations

**Comment by brawner** [comment]
Pushed some linter cleanup

File: dartsim/src/SDFFeatures.cc [hunk 2]
```diff
@@ -342,59 +372,82 @@ Identity SDFFeatures::ConstructSdfWorld(
...
-        dart::dynamics::Frame::World(),
-        _sdfModel.Name()+"_frame",
-        ResolveSdfPose(_sdfModel.SemanticPose()));
+          parentFrame, modelName + "::__model__",
+          ResolveSdfPose(_sdfModel.SemanticPose()));
 
   // Set canonical link name
   auto [modelID, modelInfo] = this->AddModel( // NOLINT
-      {model, modelFrame, _sdfModel.CanonicalLinkName()}, _worldID);
+      {model, modelFrame, _sdfModel.CanonicalLinkName()}, worldID);
```
Comment (azeey) [inline dartsim/src/SDFFeatures.cc:430]: In `ConstructSdfLink`, this canonical link name is associated with a link. I would think we need special handling of that now that the link can be nested.

Comment (brawner) [reply dartsim/src/SDFFeatures.cc:433]: After meeting we decided this feature addition would be left as a todo. I opened an issue here for tracking

https://github.com/ignitionrobotics/ign-physics/issues/209

File: dartsim/src/SDFFeatures.cc [hunk 3]
```diff
@@ -406,11 +459,68 @@ Identity SDFFeatures::ConstructSdfModel(
...
+    {
+      ignerr << "The link [" << childLinkName << "] of the child frame ["
+             << sdfJoint->ChildLinkName() << "] of joint [" << sdfJoint->Name()
+             << "] in model [" << modelName
+             << "] could not be resolved. The joint will not be constructed\n";
+      continue;
+    }
+
+    auto * const parent =
+      FindBodyNode(worlds[worldID], modelName, parentSdfLink->Name());
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: Fixed

File: dartsim/worlds/model_with_nested_model.sdf [hunk 4]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" ?>
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/worlds/model_with_nested_model.sdf]: This was part of some earlier testing. Removed

**Comment by scpeters** [comment]
potential fix for windows CI in https://github.com/ignition-tooling/gazebodistro/pull/26

**Comment by scpeters** [comment]
another prerelease of libsdformat11 should fix homebrew CI: https://github.com/osrf/sdformat/pull/488

**Comment by scpeters** [comment]
@osrf-jenkins run tests please

**Review by scpeters — COMMENTED** [review]
looks reasonable to me. I think it doesn't quite work with `//joint/parent == "world"`, so we could consider adding a test case for that

File: dartsim/src/SDFFeatures.cc [hunk 5]
```diff
@@ -342,59 +372,85 @@ Identity SDFFeatures::ConstructSdfWorld(
...
       dart::dynamics::SimpleFrame::createShared(
-        dart::dynamics::Frame::World(),
-        _sdfModel.Name()+"_frame",
-        ResolveSdfPose(_sdfModel.SemanticPose()));
+          parentFrame, modelName + "::__model__",
+          ResolveSdfPose(_sdfModel.SemanticPose()));
 
   // Set canonical link name
+  // TODO(anyone) This may work correctly with nested models and will need to be
+  // updated once more than canonical link can exist in a nested model
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: I think you're right. Cleaned up

File: dartsim/src/SDFFeatures.cc [hunk 6]
```diff
@@ -406,11 +462,68 @@ Identity SDFFeatures::ConstructSdfModel(
...
+    // Resolve parent and child frames to links
+    std::string parentLinkName;
+    ::sdf::Errors errors = sdfJoint->ResolveParentLink(parentLinkName);
+    if (!errors.empty())
+    {
+      ignerr << "The link of the parent frame [" << sdfJoint->ParentLinkName()
+             << "] of joint [" << sdfJoint->Name() << "] in model ["
+             << modelName
+             << "] could not be resolved. The joint will not be constructed\n";
+      continue;
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc:477]: Added a printout of the errors.

File: dartsim/src/SDFFeatures.cc [hunk 7]
```diff
@@ -406,11 +462,68 @@ Identity SDFFeatures::ConstructSdfModel(
...
+    }
 
-    dart::dynamics::BodyNode * const child = this->FindOrConstructLink(
-          model, modelIdentity, _sdfModel, sdfJoint->ChildLinkName());
+    const ::sdf::Link *parentSdfLink = _sdfModel.LinkByName(parentLinkName);
+    if (nullptr == parentSdfLink)
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: Fixed

File: dartsim/src/SDFFeatures.cc [hunk 8]
```diff
@@ -406,11 +462,68 @@ Identity SDFFeatures::ConstructSdfModel(
...
+             << sdfJoint->ChildLinkName() << "] of joint [" << sdfJoint->Name()
+             << "] in model [" << modelName
+             << "] could not be resolved. The joint will not be constructed\n";
+      continue;
+    }
+
+    auto * const parent =
+      FindBodyNode(this->worlds[worldID], modelName, parentSdfLink->Name());
+
+    if (nullptr == parent && parentSdfLink->Name() != "world")
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: Fixed

File: dartsim/src/SDFFeatures.cc [hunk 9]
```diff
@@ -509,11 +622,68 @@ Identity SDFFeatures::ConstructSdfJoint(
     const ::sdf::Joint &_sdfJoint)
 {
   const auto &modelInfo = *this->ReferenceInterface<ModelInfo>(_modelID);
+
+  // If we get an error here, one of two things has occured:
+  // 1. The Model _sdfJoint came from is invalid, i.e, had errors during
+  //    sdf::Root::Load and ConstructSdfJoint regardless of the errors.
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: Fixed

File: dartsim/src/SDFFeatures.cc [hunk 10]
```diff
@@ -509,11 +622,68 @@ Identity SDFFeatures::ConstructSdfJoint(
...
+  const auto childResolveErrors = _sdfJoint.ResolveChildLink(childLinkName);
+  if (!childResolveErrors.empty()) {
+    childLinkName = _sdfJoint.ChildLinkName();
+  }
 
   dart::dynamics::BodyNode * const child =
-      modelInfo.model->getBodyNode(_sdfJoint.ChildLinkName());
+    FindBodyNode(world, modelInfo.model->getName(), childLinkName);
+
+  if (nullptr == parent && _sdfJoint.ParentLinkName() != "world")
```
[the remark this answers is not in the document]
Comment (brawner) [reply dartsim/src/SDFFeatures.cc]: Fixed here and elsewhere

**Review by scpeters — APPROVED** [review]
thanks @brawner this looks good to me

I'm going to merge to simplify review of the ign-gazebo PR, but @azeey should take a look at this once he gets back to make sure he's happy with all of it

**Pull request merged**